src/urllib3/response.py

```diff
@@ -169,10 +169,15 @@ def __init__(self) -> None:
         def decompress(self, data: bytes) -> bytes:
             if not data:
                 return b""
-            return self._obj.decompress(data)  # type: ignore[no-any-return]
+            data_parts = [self._obj.decompress(data)]
```

Comment:
I'm slightly concerned about decompress performance in this case. The current design of urllib3 seems to be to just push any data it has right to zstd and ignore `zstd.DECOMPRESSION_RECOMMENDED_INPUT_SIZE`, but won't that hurt performance when downloading larger files? It's nothing new to your change, but I am wondering if there is a solution that could respect `zstd.DECOMPRESSION_RECOMMENDED_INPUT_SIZE` to not call `decompress()` so often.

Reply:
To test this I have created a 1GB file containing fully-random bytes (so that output is >1GB) and compressed it:
```
$ zstd -l data.zst
Frames  Skips  Compressed  Uncompressed  Ratio  Check  Filename
     1      0     954 MiB                       XXH64  data.zs
```

Then I run the following code with `pytest-benchmark`:

<details>
  <summary>Benchmark code</summary>
  
```python
import zstandard as zstd

with open("data.zst", "rb") as f:
    DATA = f.read()


def oneshot():
    obj = zstd.ZstdDecompressor().decompressobj()
    return len(obj.decompress(DATA))


def test_oneshot(benchmark):
    size = benchmark(oneshot)
    assert size == 1_000_000_000


def incremental():
    obj = zstd.ZstdDecompressor().decompressobj()
    size = 0
    for i in range(0, len(DATA), zstd.DECOMPRESSION_RECOMMENDED_INPUT_SIZE):
        size += len(
            obj.decompress(DATA[i : i + zstd.DECOMPRESSION_RECOMMENDED_INPUT_SIZE])
        )
    return size


def test_incremental(benchmark):
    size = benchmark(incremental)
    assert size == 1_000_000_000
```

</details>

The results show that decompressing one 1G string is ~2.5 time slower than splitting it and sending the data incrementally:

```
---------------------------------------- benchmark: 2 tests ----------------------------------------
Name (time in ms)          Min                 Max                Mean              Median          
----------------------------------------------------------------------------------------------------
test_incremental      317.5479 (1.0)      384.3817 (1.0)      358.2648 (1.0)      359.8864 (1.0)    
test_oneshot          921.7404 (2.90)     968.6641 (2.52)     949.1417 (2.65)     964.4232 (2.68)   
----------------------------------------------------------------------------------------------------
```

---

Now I leave urllib3's maintainer decide if they want me to make a quick change in the code to do so or if that's ok as it is.

One could also argue that python-zstandard should do it automatically.

Reply:
Also @grossag I did not find much documentation about `DECOMPRESSION_RECOMMENDED_INPUT_SIZE` (nor  `ZSTD_DStreamInSize()` for that matter), feel free to share if you have some poniters!

Reply:
`zstd.DECOMPRESSION_RECOMMENDED_INPUT_SIZE` defaults to the result of a call into zstd's function `ZSTD_DStreamInSize()`. After reading through some zstd source and also https://github.com/facebook/zstd/issues/340 , I think this value may represent the chunk size that zstd performs best with.

That said, optimizing for network download + decompress is much different than disk read + decompress. Also, the variance will be much higher in terms of network download speed. So it's hard to know how much to tune this.

The downside with delaying decompression until you have a buffer of size `DECOMPRESSION_RECOMMENDED_INPUT_SIZE` is that it defers more work into `flush()`, which as shown in https://github.com/urllib3/urllib3/pull/3021 can break tests. If we delay decompression, the normal case for smaller chunked transfers is that `flush()` will actually do the decompress operation.

So I am fine either way and will defer to @sethmlarson on whether more work is needed to delay decompression. Either way, I still think we should go with your approach. Mine has the benefit of using the only API that is known to be able to handle multiple frames now, which could slightly improve performance in that case. But I think the performance gain is lost by my approach having to use an intermediate `io.BytesIO()` which would result in an additional copy. I like how your approach doesn't need extra copies.

Reply:
Thank you for the pointers.

I am confused though. I though you asked about what happens when we feed **more** data than `zstd.DECOMPRESSION_RECOMMENDED_INPUT_SIZE` to the decompressor, but then in your last message you speak about what happens when we feed **less** data. So here are my thoughts:

- Feeding more data: we see in the benchmark that there seems to be a penalty cost to do so (600ms per GB). I'm open to changing the code to consider that but would argue that the python-zstandard's lib should do it itself.
- Feeding less data: this case allows for the consumer of urllib3 to get the data as soon as it is ready, which would be helpful when streaming HTTP responses. I'm not sure how often this scenario would happen in real life though.

Also, you spoke about calling `flush()`, but in the case of this PR `ZstdDecompressionObj.flush` does nothing at all. I put a comment in the code for clarity, but did not remove the code calling it in case python-zstandard's API changes in the future.

Reply:
Sorry! I am wondering specifically about feeding less data than `zstd.DECOMPRESSION_RECOMMENDED_INPUT_SIZE`. But as discussed in Discord, zstd frames can be any size so my concern isn't important.

My biggest concern was: what happens when we receive half of a frame's worth of data? python-zstandard's underlying call to zstd's `ZSTD_decompressStream` will return 3, meaning that there is more data to decode. But there isn't enough data to decode so we basically need to wait for more data.

I wrote some Python code where I read the full contents of a .zstd file and sliced it in half at different points. I then ran your code over it. It worked perfectly for various ways that I sliced the buffer, so I think your approach is great.

The part I didn't realize is that zstd internally remembers where it is at during a decompress operation. If there is not enough data, it will do as much as it can and return. Then your code is doing the right thing in reusing `self._obj` if it doesn't have enough data and it's mid-frame (`self._obj.eof` being False if we are mid-frame).

Reply:
Great! Thank you for taking the time to confirming that everything works as expected! :+1: